Ensure cs_utxo is locked by the caller when a coin reference is returned

ptschip opened the pull request:
Thanks to @dagurval for catching this bug.

AccessCoin() and AccessByTxid() both return references to a coin in the dbcache.  The caller of those functions has to make sure the cs_utxo locked during the time the returned coin reference is being used, so by using AssertLockHeld() at the beginning of both AccessCoin() and AccessByTxid() we force the caller to to take the lock where/when appropriate.

ptschip committed:
Fix locking of AccessCoin() and AccssByTxid()

Both these functions return a reference to a coin in the coins cache
and as a result we have to make sure the caller has locked cs_utxo
for as long as it uses the returned reference.

ptschip committed:
Ensure that cs_utxo is locked in CheckTxInputs()

When we call AccessCoin() we must make sure that the returned coin
reference is still locked for as long as we need it.

ptschip committed:
fix format

sickpig reviewed (approved):
good catch @dagurval

gandrewstone commented on src/consensus/tx_verify.cpp:
```diff
@@ -212,26 +215,48 @@ bool Consensus::CheckTxInputs(const CTransaction &tx, CValidationState &state, c
+                {
+                    // GetSpendHeight requires cs_main but in order to maintain proper locking order cs_utxo
+                    // must be taken after cs_main.
+                    LEAVE_CRITICAL_SECTION(inputs.cs_utxo);
```
Leave and re-enter is a bit of a lie because the protected variables cannot be used after inputs.cs_utxo is given, even if it is relocked.  So I would like to discourage its use, since it will be prone to bugs.  In this case, the code correctly does not use "coin" after the lock is given.  

However, this code would be a lot more clear if the lock just protected pulling the relevant data out of "coin" and into local variables, at the top of the for loop body.  This is already basically happening in lines 229 and 230, so a bit of reorg (pull these locals out of the if, assign them, and use in the else as well) could make this code both cleaner and lock cs_utxo for less time.

ptschip replied:
We could do the copy for all checks and that would be cleaner, but here I was trying to avoid the copy on most iterations because coinbase inputs are relatively rare.

gandrewstone commented on src/main.cpp:
```diff
@@ -1396,7 +1402,7 @@ bool CheckInputs(const CTransaction &tx,
                 // spent being checked as a part of CScriptCheck.
                 const CScript &scriptPubKey = coin.out.scriptPubKey;
-                const CAmount amount = coin.out.nValue;
+                const CAmount &amount = coin.out.nValue;
```
similarly here, if we copy the script and amount, we'll be able to release the lock, and let coin fall out of scope.  This means that the lock won't be held during the script execution which will allow us to do it in parallel.  Since script execution generally contains sig validations, its pretty certain that for almost all scripts the copy time is trivial compared to script execution.

CScript scriptPubKey;
CAmount amount;
{
LOCK(inputs.cs_utxo);
￼const Coin &coin = inputs.AccessCoin(prevout);		￼                  
￼if (coin.IsSpent()) LOGA("ASSERTION: no inputs available\n");		￼                      assert(!coin.IsSpent());
scriptPubKey = coin.out.scriptPubKey;
amount = coin.out.nValue;
}

ptschip replied:
yeah, i was trying to avoid the copy but you're right, it's not worth it.  i'll update it...

ptschip committed:
Temporarily lock cs_utxo while we copy the amount and scriptPubKey

This way we minimize the time cs_utxo is locked and prevent it from
being locked while the scripts and sigs are being checked.

ptschip committed:
Add clarifying comment as to why we do the copy

We need the copy becasue we can't depend on "coin" still being
valid.

We could do the copy further up and then release the lock before
taking cs_main however that would require a copy each and every
time we CheckTxInpts() when in reality this section of the code
gets executed rarely.
